python/lsst/pipe/base/graphBuilder.py

```diff
@@ -28,36 +29,674 @@
+        # Note that there's only one scaffolding node for each DatasetType, shared by
+        # _PipelineScaffolding and all _TaskScaffoldings that reference it.
+        self.tasks = [_TaskScaffolding(taskDef=taskDef, parent=self, datasetTypes=taskDatasetTypes)
+                      for taskDef, taskDatasetTypes in zip(pipeline, datasetTypes.byTask.values())]
+
+    tasks: List[_TaskScaffolding]
+    """Scaffolding data structures for each task in the pipeline
+    (`list` of `_TaskScaffolding`).
+    """
+
+    initInputs: _DatasetScaffoldingDict
+    """Datasets consumed but not produced when constructing the tasks in this
+    pipeline (`_DatasetScaffoldingDict`).
+    """
+
+    initIntermediates: _DatasetScaffoldingDict
+    """Datasets that are both consumed and produced when constructing the tasks
+    in this pipeline (`_DatasetScaffoldingDict`).
+    """
+
+    initOutputs: _DatasetScaffoldingDict
+    """Datasets produced but not consumed when constructing the tasks in this
+    pipeline (`_DatasetScaffoldingDict`).
+    """
+
+    inputs: _DatasetScaffoldingDict
+    """Datasets that are consumed but not produced when running this pipeline
+    (`_DatasetScaffoldingDict`).
+    """
+
+    intermediates: _DatasetScaffoldingDict
+    """Datasets that are both produced and consumed when running this pipeline
+    (`_DatasetScaffoldingDict`).
+    """
+
+    outputs: _DatasetScaffoldingDict
+    """Datasets produced but not consumed when when running this pipeline
+    (`_DatasetScaffoldingDict`).
+    """
+
+    prerequisites: _DatasetScaffoldingDict
+    """Datasets that are consumed when running this pipeline and looked up
+    per-Quantum when generating the graph (`_DatasetScaffoldingDict`).
+    """
+
+    dimensions: DimensionGraph
+    """All dimensions used by any regular input, intermediate, or output
+    (not prerequisite) dataset; the set of dimension used in the "Big Join
+    Query" (`DimensionGraph`).
+
+    This is required to be a superset of all task quantum dimensions.
+    """
+
+    def fillDataIds(self, registry, originInfo, userQuery):
+        """Query for the data IDs that connect nodes in the `QuantumGraph`.
+
+        This method populates `_TaskScaffolding.dataIds` and
+        `_DatasetScaffolding.dataIds` (except for those in `prerequisites`).
+
+        Parameters
+        ----------
+        registry : `lsst.daf.butler.Registry`
+            Registry for the data repository; used for all data ID queries.
+        originInfo : `lsst.daf.butler.DatasetOriginInfo`
+            Object holding the input and output collections for each
+            `DatasetType`.
+        userQuery : `str`, optional
+            User-provided expression to limit the data IDs processed.
+        """
+        # Initialization datasets always have empty data IDs.
+        emptyDataId = DataId(dimensions=registry.dimensions.empty)
+        for datasetType, scaffolding in itertools.chain(self.initInputs.items(),
```

Comment:
`datasetType` is not used by the loop, you can probably iterate over `values()`.